### PR TITLE
Enable Vanilla Bwd and Refactor

### PR DESCRIPTION
**Pull request by micmelesse**
This PR 
* enables vanilla bwd
* refactors our code into several files. 
* fixes a bug with softmax lse that was returned with forward
* adds an alternate mode that uses `exp` instead of `exp2`. This was useful to debug issues with both forward and backward
* creates interfaces for fa and pytorch that use implementation functions with explicit paramaters.
* adds a pytorch ref implementations that mimics our triton kernels for testing.

### CODE REVIEW

**Comment by micmelesse** [comment]
The tests pass on MI300 but seems the ci MI200 have issues.
image
